clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -2253,7 +2253,9 @@ class SyclKernelArgsSizeChecker : public SyclKernelFieldHandler {
     const CXXRecordDecl *RecordDecl = FieldTy->getAsCXXRecordDecl();
     assert(RecordDecl && "The type must be a RecordDecl");
     llvm::StringLiteral MethodName =
-        IsSIMD ? InitESIMDMethodName : InitMethodName;
+        (IsSIMD && Util::isSyclType(FieldTy, "accessor", true /*Tmp*/))
```

Comment:
So, the reason for crash is that in ESIMD mode we search for `__init_esimd` and our implementation of sampler/stream don't have them.
I wonder if it is more user friendly to emit some error that "sampler/stream" is not supported in ESIMD context instead of silently trying to initialize sampler/stream with default `__init` methods.
@kbobrovs , maybe you can comment on that?

Reply:
"So, the reason for crash is that in ESIMD mode we search for __init_esimd and our implementation of sampler/stream don't have them."
Yes that's correct.
"I wonder if it is more user friendly to emit some error that "sampler/stream" is not supported in ESIMD context instead of silently trying to initialize sampler/stream with default __init (non-ESIMD) methods."
Yes that's a good idea. I can add a diagnostic here and in the other places in SemaSycl.cpp using the same condition about the MethodName.

Reply:
IIRC, when implementing PR for `sycl_special_attribute`, we discussed this issue (https://github.com/intel/llvm/pull/3892/files/d7a510563fcbd820882251a06ea471d2d666364e#r705702417). This was a known existing problem we did not fix in that PR because it was existing behavior (it just crashed in `ArgChecker `prior to https://github.com/intel/llvm/pull/3892).

Based on that thread, I think this issue requires more analysis. We can probably add a diagnostic as anything is better than a crash, but based on discussion here https://github.com/intel/llvm/pull/3892/files/d7a510563fcbd820882251a06ea471d2d666364e#r707236082, we expect to get ride of __init_esimd.

Reply:
With the diagnostic, I think we may not need this change anymore.

Reply:
Yes, eventually we will want to support stream and (less likely) sampler, and we want to get rid of __init_esimd in favor generic init + dead arguments elimination.

Reply:
Thanks for clarifying @kbobrovs!